File: cubist-node-sdk/src/config/axelar_manifest.ts
import * as fs from 'fs';
import { validateAxelarManifest, } from './schema/validator';
import { PathBuf, } from '../config';
import { Address, } from '../internal/target_handler/solidity';

/**
 * Per-target manifest file that the Axelar relayer produces.
 *
 * @group Internal
 */
export interface IAxelarManifest {
    /** Chain name */
    name: string,
    /** Chain id */
    chainId: number,
    /** Gateway contract address */
    gateway: Address,
    /** Gas receiver contract address */
    gasReceiver: Address,
    /** Deployer contract address */
    constAddressDeployer: Address,
}

/**
 * Per-target manifest file that the Axelar relayer produces.
 *
 * @group Internal
 */
export class AxelarManifest {
  private _name: string;
  private _chainId: number;
  private _gateway: string;
  private _gasReceiver: string;
  private _constAddressDeployer: string;

  /** Chain target */
  get name(): string {
    return this._name;
  }

  /** Chain id */
  get chainId(): number {
    return this._chainId;
  }

  /** Gateway contract address */
  get gateway(): Address {
    return this._gateway;
  }

  /** Gas receiver contract address */
  get gasReceiver(): Address {
    return this._gasReceiver;
  }

  /** Deployer contract address */
  get constAddressDeployer(): Address {
    return this._constAddressDeployer;
  }

  /** @ignore Empty constructor */
  private constructor() {
    // @eslint-disable-line @typescript-eslint/no-empty-function
  }

  /**
   * Create manifest from JSON file.
   * @param {PathBuf} file Path to the manifest file.
   * @return {AxelarManifest} the manifest.
   */
  static from_file(file: PathBuf): AxelarManifest {
    const json = JSON.parse(fs.readFileSync(file, 'utf8'));
    return AxelarManifest._from_json(json);
  }

  /**
   * Create manifest from JSON object.
   *
   * @param {IAxelarManifest} json the manifest object.
   * @return {AxelarManifest} the manifest.
   * @internal
   */
  static _from_json(json: IAxelarManifest): AxelarManifest {
    const self = new AxelarManifest();

    // Validate the json against the schema
    validateAxelarManifest(json);

    // Populate the object
    self._name = json.name;
    self._chainId = json.chainId;
    self._constAddressDeployer = json.constAddressDeployer;
    self._gasReceiver = json.gasReceiver;
    self._gateway = json.gateway;

    return self;
  }
}

File: cubist-config/src/axelar_manifest.rs
use std::fmt::{Debug, Display};

use ethers_core::types::Address;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// Axelar chain names, taken from
///  - `<https://docs.axelar.dev/dev/build/chain-names/mainnet>`
///  - `<https://docs.axelar.dev/dev/build/chain-names/testnet>`
#[derive(Serialize, Deserialize, Clone, JsonSchema)]
#[allow(non_camel_case_types)]
#[allow(missing_docs)]
pub enum ChainName {
    acre,
    agoric,
    arbitrum,
    assetmantle,
    aura,
    aurora,
    Avalanche,
    Axelarnet,
    binance,
    burnt,
    celo,
    comdex,
    #[serde(rename = "comdex-2")]
    comdex_2,
    cosmoshub,
    crescent,
    #[serde(rename = "e-money")]
    e_money,
    Ethereum,
    #[serde(rename = "ethereum-2")]
    ethereum_2,
    evmos,
    Fantom,
    fetch,
    injective,
    juno,
    kava,
    ki,
    kujira,
    Moonbeam,
    optimism,
    osmosis,
    #[serde(rename = "osmosis-5")]
    osmosis_5,
    persistence,
    Polygon,
    regen,
    secret,
    sei,
    stargaze,
    terra,
    #[serde(rename = "tera-2")]
    terra_2,
    #[serde(rename = "tera-3")]
    terra_3,
    umee,
    xpla,
}

fn to_str(cn: &ChainName) -> String {
    serde_json::to_string(cn)
        .unwrap()
        .trim_matches('"')
        .to_owned()
}

impl Display for ChainName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", to_str(self))
    }
}

impl Debug for ChainName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", to_str(self))
    }
}

/// Per-target manifest file that the Axelar relayer produces
#[derive(Serialize, Deserialize, Clone, JsonSchema)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive] // because generated json file may have other fields, e.g., 'tokens'
pub struct AxelarManifest {
    /// Chain name
    pub name: String,
    /// Chain id
    pub chain_id: u32,
    /// Gateway contract address
    #[schemars(with = "String")]
    pub gateway: Address,
    /// Gas receiver contract address
    #[schemars(with = "String")]
    pub gas_receiver: Address,
    /// Deployer contract address
    #[schemars(with = "String")]
    pub const_address_deployer: Address,
}
